### PR TITLE
feat: inline-rename a route heatmap region from its detail page

### PR DESCRIPTION
**Pull request by llun**
## Summary

Brings the route-heatmap **detail page** in line with the design system kit (`ui_kits/web/heatmaps.html` → `RegionHeatmapDetail.jsx`'s `EditableRegionName`).

A drawn area's title on its own heatmap page is now an **inline-editable name**:

- Click the title (or the pencil that appears on hover) to edit it in place — no popup.
- **Enter** or **blur** saves; **Escape** cancels; an empty or unchanged value just closes the editor and keeps the existing name.
- The **whole-world** region and any read-only caller (no `onRename`) keep a static title.

This matches the design's intent that each region is renamed from its own page. The existing picker **pencil / "Edit area" composer is unchanged**, so editing a rectangle's bounds still works there.

## How it's wired

- `RegionHeatmapDetail` gains an optional `onRename(region, name)` prop and renders the new internal `EditableRegionName` for drawn areas. The title stays an `<h2>` so the page-level `PageHeader` keeps the only `<h1>`.
- `FitnessHeatmapView` passes `onRename={handleRegionRenamed}`, which:
  - updates the in-memory region list immediately (the open detail page reads its title from there, so the heading updates without a refetch), and
  - persists the label by **canonical region key** via the existing `setFitnessRouteHeatmapRegionName` path. The key is **coordinate-only** (`serializeRegion` ignores `name`), so renaming never changes the heatmap link or its cached row.

## Tests

- `RegionHeatmapDetail.test.tsx`: read-only world title, static title without `onRename`, commit on Enter, commit on blur, cancel on Escape, and no-op on an unchanged name.
- `FitnessHeatmapView.test.tsx`: opening a seeded drawn region, renaming it inline, and asserting the label persists against the rect key while the heading updates in place.

## Verification

- `yarn lint` — 0 errors
- `yarn build` — passes
- `yarn test` — 549 files / 4840 tests pass

No migrations or config changes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by llun** [comment]
## 🤖 Sub-agent review — round 1

Reviewed the diff (`RegionHeatmapDetail.tsx` + `EditableRegionName`, `FitnessHeatmapView.handleRegionRenamed`, and both test files) against the surrounding code. **No blocking issues found** — the edit/commit/cancel logic, persistence keying, state sync, accessibility, hooks ordering, and project conventions all check out.

### Blocking
- None.

### What I verified (all correct)
- **Idempotent commit guards the double-fire cases.** `EditableRegionName.commit` only fires `onRename` when `next && next !== (value ?? '')` (`RegionHeatmapDetail.tsx:359-365`). So the classic inline-edit hazards are safe: Enter → `commit()` → `setEditing(false)` re-renders the parent with the new `value`, so any stray unmount `blur` re-runs `commit` as a no-op (`next === value`). Escape → `cancel()` resets `draft` to `value` first, so a stray commit is also a no-op. Tests confirm `toHaveBeenCalledTimes(1)` on Enter and 0 on Escape/unchanged.
- **Rename keeps the heatmap link intact.** `serializeRegion`/`rectToken` (`lib/fitness/regions.ts:80-115`) read only `nw`/`se` — `name` is explicitly UI-only (`RectRegion.name`, line 32-33). Persisting via `handleRegionSaved` uses the coordinate-only key, so the rename can't move the cache row or fight the dedupe path. `name` length is consistent end-to-end: input `maxLength={80}` < server `z.string().max(255)` (route `…/fitness-route-heatmap-region-names/route.ts:39-44`).
- **Heading updates without a refetch, no stale closure.** `openRegion` is `useMemo(() => regions.find(r => r.id === openRegionId), …)` (`FitnessHeatmapView.tsx:259-261`); `handleRegionRenamed` updates the matching `regions` entry by `id` with a functional updater and depends only on `[handleRegionSaved]` (lines 699-720), so the open detail page re-derives its title from the new state. Integration test asserts exactly this.
- **Hooks/Next correctness.** All hooks (incl. `handleRegionRenamed`) sit above the `if (openRegion) return …` early return (`FitnessHeatmapView.tsx:723`); `EditableRegionName` is its own component so its `useState`/`useEffect`/`useRef` are unconditional. No new render-time `Date.now()` introduced.
- **A11y.** Edit affordance is a real `<button>` whose accessible name comes from the visible `<h2>` text (`title="Rename"` does not override visible text), and the input carries `aria-label="Region name"` with focus+select on entry. World region / no-`onRename` callers keep a static `<h2>`.
- **Conventions.** No `console.*` or direct `fetch` in the component; the API call goes through `setFitnessRouteHeatmapRegionName` in `lib/client.ts`; design-system tokens used throughout; tests use plain descriptive `it(...)` names.

### Suggestions (optional, non-blocking)
- **Two uncovered edge cases** the PR description calls out but doesn't test: (a) **blur immediately after Escape**, and (b) **rename → back → reopen** (in-session persistence of the new name via `regions`). Both work by construction, but a regression test for each would lock the behavior in.

### Nits
- `RegionHeatmapDetail.tsx:513` `value={region.type === 'rect' ? region.name : undefined}` — the ternary is dead/defensive: this branch only renders when `!isWorld && onRename`, so `region` is always `rect` here. It's a harmless TS narrowing helper; could be simplified to `region.name` but not worth a revision.
- During editing the title `<h2>` is replaced by the `<input>`, so the heading momentarily leaves the document outline. Semantically fine (the input is the title being edited); noting only for completeness.

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 1]
```diff
@@ -321,6 +322,92 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
...
+        onBlur={commit}
+        onKeyDown={(event) => {
+          if (event.key === 'Enter') {
+            event.preventDefault()
+            commit()
+          } else if (event.key === 'Escape') {
+            event.preventDefault()
+            cancel()
+          }
+        }}
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx:410]: Fixed in 99777c8d6. Closing the editor now flows through a single `close(save)` path guarded by a `closingRef`, so the blur that some browsers fire when the focused input unmounts is a no-op — Enter no longer double-commits and a trailing blur can't override Escape. I used a *reset-on-open* guard rather than resetting only inside `commit()`: the latter leaves the ref stuck `true` in environments that don't fire blur-on-unmount (e.g. jsdom), which would silently swallow the next edit's save. Clearing the name is now supported too (sibling threads).

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 2]
```diff
@@ -321,6 +322,92 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
   )
 }
 
+interface EditableRegionNameProps {
+  /** Current area name; falls back to the generic "Map area" placeholder. */
+  value?: string
+  /** Fired with the trimmed, changed, non-empty name when an edit commits. */
+  onRename: (name: string) => void
+}
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx:334]: Done in 99777c8d6 — the `EditableRegionName` interface doc now states that an empty string clears the name, and `commit` persists it. This matches the picker's edit-area composer and the backend, which already support clearing a label (`setFitnessRouteHeatmapRegionName(..., null)`).

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 3]
```diff
@@ -321,6 +322,92 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
...
+    }
+  }, [editing])
+
+  const commit = () => {
+    const next = draft.trim()
+    // Only persist a real, non-empty change — an empty or unchanged value just
+    // closes the editor and keeps the existing name.
+    if (next && next !== (value ?? '')) onRename(next)
+    setEditing(false)
+  }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx]: Adopted in 99777c8d6: the guard is now `if (next !== (value ?? '')) onRename(next)`, so an emptied field clears the name; `handleRegionRenamed` in the view maps the empty string to a cleared label. Combined with the reentrancy guard from the other thread.

File: lib/components/fitness/RegionHeatmapDetail.test.tsx [hunk 4]
```diff
@@ -337,4 +337,111 @@ describe('RegionHeatmapDetail', () => {
...
+        onRename={onRename}
+      />
+    )
+    fireEvent.click(screen.getByRole('button', { name: /Veluwe loop/i }))
+    fireEvent.keyDown(screen.getByRole('textbox', { name: 'Region name' }), {
+      key: 'Enter'
+    })
+    expect(onRename).not.toHaveBeenCalled()
+  })
 })
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.test.tsx:590]: Added in 99777c8d6 — `clears the name when an empty (whitespace-only) value is committed` asserts `onRename(region, '')`. Also added in-session persistence tests (back/reopen, and editor-reopen reflecting an external rename).

**Comment by llun** [comment]
### Round 1 addressed (commit 99777c8d6)

Thanks @gemini-code-assist and the sub-agent reviewer. Changes:

- **Idempotent editor close** — a single `close(save)` path guarded by a reset-on-open `closingRef` absorbs the unmount-blur, so Enter can't double-commit and a trailing blur can't override Escape. (Chose reset-on-open over a commit-only reset so the guard can't poison a later edit when no blur-on-unmount fires.)
- **Clearing a name** — an empty commit now reverts to the "Map area" placeholder, matching the picker's edit-area composer and the backend.
- **Tests** — clear-on-empty, plus in-session persistence across back/reopen and editor reopen.

All four review threads resolved. `yarn lint` / `yarn build` / `yarn test` (4843) green.

**Comment by llun** [comment]
## 🤖 Sub-agent review — round 2

Re-scrutinized the latest commit (`99777c8d6` "fix: harden inline heatmap-region rename per review"). The two round-1 fixes (idempotent `close(save)` + clear-on-empty) are correct and well-covered. **Nothing blocking.** Ran both changed test files (74 passed) and lint (clean) before reporting.

### Blocking
None.

### Verified (no issues)
- **`closingRef` guard is correct.** `close()` sets `closingRef.current = true` *before* `setEditing(false)` (`RegionHeatmapDetail.tsx:367-378`), so the unmount-blur that follows hits the early `return`. The ref only resets to `false` inside the `useEffect([editing])` when `editing === true` (`:355`), and `close()` is only reachable from the input's `onBlur`/`onKeyDown` — which can't fire while the read-only button is mounted. So there is no path where `closingRef` stays `true` and silently blocks a later save, and no path where the unmount-blur double-commits (Enter) or resurrects a cancelled edit (Escape). Commit logic runs exactly once for Enter, blur, and Escape.
- **Clearing is key-safe end-to-end.** `serializeRegion`/`rectToken` (`lib/fitness/regions.ts:80-115`) key only on coordinates — `name` is explicitly "Not part of the serialized cache key" (`regions.ts:32`). Empty commit → `onRename(region,'')` → `handleRegionRenamed` (`FitnessHeatmapView.tsx:699-717`): `trimmed = '' || undefined`, `setRegions` matches by client `id` (not the canonical key) and sets `name: undefined` (heading → "Map area"), then `handleRegionSaved` serializes the unchanged coordinate key and PUTs `name: null` (`client.ts:2566-2575`). The region's canonical key and heatmap link cannot change on rename/clear. Confirmed `openRegion` re-memoizes with the new `name` but the same `id`, so `openRegionKey`/`focusKey` stay stable (no detach).
- **No accidental clear.** Opening pre-selects the text (`useEffect([editing])` → `select()`), and `close(true)` no-ops when `next === (value ?? '')` (`:374`). The "does not save an unchanged inline rename" test confirms an untouched Enter does not fire `onRename`.
- **Hooks/closures/hydration sound.** All `useState`/`useRef`/`useEffect` precede the conditional `if (editing) return` — no conditional hooks. `useEffect([value, editing])` guards on `!editing`, so an external `value` change can't clobber a draft mid-edit and resyncs on close (matches the "reflects an externally-updated name when the editor reopens" test). `handleRegionRenamed` deps include `handleRegionSaved` — no stale closure. No `Date.now()`/`new Date()` in render.
- **Conventions.** No `console.*`, no direct `fetch()` (goes through `lib/client.ts`), design tokens only (`border-input`/`ring-ring`/`bg-accent`), `maxLength={80}` + `trim() || undefined` match the picker's edit-area composer (`HeatmapRegionPicker.tsx:234,319`), plain test names. `onRename` is optional and the sole caller is `FitnessHeatmapView` — backward compatible.

### Nits (non-blocking)
- No test asserts the *detail page's* editable heading flips back to "Map area" within the same session after a clear, nor that a nameless rect renders the "Map area" affordance in the detail view. The clear path is covered at the unit level (`onRename(region,'')`) and the persistence/list path at the integration level, and "Map area" is covered in the view tests — so this is a small coverage gap, not a correctness risk.

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 5]
```diff
@@ -321,6 +322,106 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
...
+      onClick={() => setEditing(true)}
+      title="Rename"
+      className="group/name -mx-1.5 inline-flex max-w-full items-center gap-1.5 rounded-md px-1.5 py-0.5 text-left transition-colors hover:bg-accent focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring"
+    >
+      <h2 className="truncate text-xl font-semibold tracking-tight">
+        {value || 'Map area'}
+      </h2>
+      <Pencil className="size-3.5 shrink-0 text-muted-foreground opacity-0 transition-opacity group-hover/name:opacity-100" />
+    </button>
+  )
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx:431]: Fixed in d7c486619 — inverted the nesting so the `<h2>` wraps the `<button>` (button now contains only a `<span>` + the pencil icon, i.e. phrasing content). The heading stays in the document outline and the visual styles/truncation are preserved.

**Comment by llun** [comment]
### Round 2 addressed (commit d7c486619)

- **Valid heading markup** (@gemini-code-assist) — the `<h2>` now wraps the `<button>` (button contains only phrasing content), keeping the heading in the document outline.
- **Coverage** (sub-agent nit) — added tests for the nameless-area editable "Map area" placeholder and the heading reverting to it after a clear.

Review thread resolved. `yarn lint` / `yarn build` / `yarn test` (4845) green.

**Comment by llun** [comment]
## 🤖 Sub-agent review — round 3

Re-reviewed the current branch (`feat/heatmap-detail-inline-rename`, through `d7c4866`). The round-2 heading-markup fix is solid and the rename/clear/idempotent-close logic all holds up. **Nothing blocking — this looks ready to merge.** Verified `yarn vitest run` on both changed test files (76 passed) and `yarn lint` (0 errors, no warnings in the changed files).

### Verification of the items flagged for this round
- **Valid + accessible heading markup** (`RegionHeatmapDetail.tsx:412-431`): `<h2 class="truncate"><button title="Rename"><span class="truncate">…</span><Pencil/></button></h2>` is valid (button holds only phrasing content) and the heading stays in the outline. Confirmed the editable title is queryable as `getByRole('heading', { level: 2, name: 'Veluwe loop' })` — accessible name comes from the button's text content; the Lucide `<Pencil>` is auto-`aria-hidden` (lucide-react adds `aria-hidden="true"` when it has no children/a11y prop) and `title="Rename"` does not override name-from-content, so there's no duplicate/confusing naming. Truncation works (`h2.truncate` + inner `span.truncate` inside the `inline-flex max-w-full` button). 👍
- **Idempotent close** (`RegionHeatmapDetail.tsx:373-413`): `close(save)` + `closingRef` fire commit exactly once for Enter / blur / Escape. Enter→`preventDefault`→`close(true)` then the unmount-blur `close(true)` early-returns; Escape→`close(false)` and the unmount-blur can't resurrect a save; reset-on-open (`closingRef.current=false` in the `[editing]` effect) prevents a stale guard from poisoning the next edit (incl. the jsdom no-blur case). Markup change didn't disturb any of this.
- **Clearing can't change the key / detach the heatmap**: empty commit → `onRename(region, '')` → `handleRegionRenamed` sets `name: undefined` and calls `handleRegionSaved` with `name: null`. The canonical key is `serializeRegion(toHeatmapRegion(region))` which is coordinate-only (`serializeRegions` ignores `name`), and `openRegionId` is keyed by the stable `region.id`, so the open heatmap link is unaffected. End-to-end clearing is covered by both component- and view-level tests.
- **View handler** (`FitnessHeatmapView.tsx:699-717`): in-memory `setRegions` (so the open detail title updates without a refetch) + persistence via the picker's `handleRegionSaved` (which surfaces save failures). Hooks ordering, deps (`[handleRegionSaved]`), and the no-stale-closure path are sound; the `region.type !== 'rect'` guard is correct (world is never renamed).
- **Conventions**: no `console.*`, no direct `fetch()` (persists via `setFitnessRouteHeatmapRegionName` in `lib/client.ts`), `maxLength={80}` matches the picker's edit-area composer, design tokens throughout, plain descriptive test names.

### Suggestions
_None._

### Nits (non-blocking, optional)
- The new editable-title tests assert via `getByRole('button', …)` only; none assert the editable title is still a level-2 `heading` (the read-only/world tests do). Since valid heading markup was the whole point of `d7c4866`, a one-line `getByRole('heading', { level: 2, name: 'Veluwe loop' })` assertion in the editable case would lock that in against regressions. (I verified it passes locally.) — `RegionHeatmapDetail.test.tsx:362-401`

LGTM. ✅

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 6]
```diff
@@ -321,6 +322,109 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
...
+  // the heading in the document outline while the button stays the affordance.
+  return (
+    <h2 className="truncate text-xl font-semibold tracking-tight">
+      <button
+        type="button"
+        onClick={() => setEditing(true)}
+        title="Rename"
+        className="group/name -mx-1.5 inline-flex max-w-full items-center gap-1.5 rounded-md px-1.5 py-0.5 text-left align-middle transition-colors hover:bg-accent focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring"
+      >
+        <span className="truncate">{value || 'Map area'}</span>
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx]: Fixed in 07816a217 — added `min-w-0` to the truncating `<span>` so it can shrink below its content inside the inline-flex button; a long name now ellipses instead of pushing the pencil out.

**Comment by llun** [comment]
### Round 3 addressed (commit 07816a217)

- **Flex truncation** (@gemini-code-assist) — added `min-w-0` to the truncating `<span>` so a long region name ellipses inside the inline-flex edit button instead of pushing the pencil out.
- **Regression test** (sub-agent) — assert the editable title remains a level-2 heading.

Thread resolved. `yarn lint` / `yarn build` / `yarn test` (4846) green.

**Comment by llun** [comment]
## 🤖 Sub-agent review — round 4

Fresh skeptical pass over the current branch state (4 commits, `RegionHeatmapDetail.tsx` + `FitnessHeatmapView.tsx` + their two test files). **Nothing new found — converged and merge-ready.**

**Blocking:** none.

**Suggestions:** none.

**Nits:** none.

What I re-verified end-to-end (not just re-reading the round-3 fixes):

- **Key stability — the core correctness question.** `name` is documented and treated as UI-only and is explicitly *not* part of the serialized cache key (`lib/fitness/regions.ts:32-33`; `serializeRegion`/`serializeRegions` ignore it). So a rename leaves `serializeRegion(toHeatmapRegion(openRegion))` unchanged → `openRegionKey`, `focusKey`, the heatmap link, and the detail's `heatmapData` are all unaffected. `handleRegionRenamed` updates the matching entry in `regions` by stable `id` (`FitnessHeatmapView.tsx:707-711`); `openRegion` is `regions.find(id===openRegionId)`, so the heading re-renders with the new `value` without a remount or refetch. Sound.
- **No real concurrency picker↔detail.** The picker only renders in the master view; `if (openRegion) return <RegionHeatmapDetail …>` (`:723`) means the detail and picker are never mounted together, so detail-edit vs picker-edit of the same region can't race. Both funnel persistence through `handleRegionSaved` → `setFitnessRouteHeatmapRegionName` (same path as #1073).
- **Rename during a generation in flight** only touches `name`; `heatmapData` is keyed on `focusKey` (independent of name). Safe. Rename-then-navigate-back: `close()` commits synchronously via `onRename`, and `handleRegionSaved` is fired with `void` (`:714`) — the back button just flips `openRegionId`; the in-flight save still resolves and only sets/clears `error` on the still-mounted view. No stale-state write to an unmounted child.
- **Clear / no-op semantics** are internally consistent: `close(save)` compares `draft.trim() !== (value ?? '')`, so an unchanged name (incl. an already-nameless region committed empty) is a true no-op, while clearing fires `onRename('')` → `handleRegionRenamed` maps `'' → undefined` → `name: undefined` → heading reverts to the `Map area` placeholder. Stored names are always pre-trimmed (this editor + picker `HeatmapRegionPicker.tsx:319`), so the only theoretical spurious-save (open+Enter on a whitespace-padded stored name) can't occur in practice.
- **`closingRef` idempotency** still holds: reset on every editor-open effect (`:364`), so a single Enter/blur/Escape commits exactly once and a stray unmount-blur can neither double-commit nor resurrect a cancelled edit; it can't poison a later edit if no such blur fires.
- **Accessibility/markup:** the editable title is `<h2><button>` (button as the affordance inside the heading — valid, button contains only phrasing content), keeping one `<h2>` in the outline under the page `<h1>`; world/read-only callers keep a static `<h2>`. `maxLength={80}` matches the picker.
- **Tests:** not brittle — they assert behavior via roles (`heading level 2`, `textbox "Region name"`, `Open … heatmap`) and cover Enter/blur/Escape, no-op, clear→placeholder, externally-updated value on reopen, world stays read-only, and the view-level persist-by-coordinate-key + reopen-keeps-name paths. No false confidence spotted.

**Verification run locally:**
- `yarn vitest run` on both changed test files → **77 passed**.
- `yarn lint` → **0 errors** (496 pre-existing warnings, none in either changed source file).

Per project conventions (AGENTS.md): no direct `fetch` in components (goes through `lib/client.ts`), no `new Date()`/`Date.now()` in render, `<h1>` ownership preserved. Recommend merge.

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 7]
```diff
@@ -321,6 +322,109 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
...
+        aria-label="Region name"
+        value={draft}
+        maxLength={80}
+        placeholder="Map area"
+        onChange={(event) => setDraft(event.target.value)}
+        onBlur={() => close(true)}
+        onKeyDown={(event) => {
+          if (event.key === 'Enter') {
+            event.preventDefault()
+            close(true)
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx:405]: Fixed in c3656a74d — the editor's `onKeyDown` now early-returns when `event.nativeEvent.isComposing` is true, so during an IME composition (Japanese / Chinese / Korean) Enter confirms the candidate and Escape cancels the composition without closing or committing the editor. I read `isComposing` off the native event since React's synthetic `KeyboardEvent` type doesn't expose it, and guarded both Enter and Escape (Escape cancels the in-progress composition during compose). Added a regression test (`ignores Enter while an IME composition is active`).

**Comment by llun** [comment]
### Round 4 addressed (commit c3656a74d)

- **IME composition** (@gemini-code-assist) — the editor's keydown now early-returns on `event.nativeEvent.isComposing`, so Enter/Escape during a Japanese/Chinese/Korean composition no longer prematurely commit/close the editor. Added a regression test.

Thread resolved. Sub-agent round 4 reported converged/merge-ready. `yarn lint` / `yarn build` / `yarn test` (4847) green.

**Comment by llun** [comment]
## 🤖 Sub-agent review — round 5

Genuinely skeptical re-pass over the full `main...feat/heatmap-detail-inline-rename` diff (5 commits), the wiring in `FitnessHeatmapView.tsx`, the backend `fitness-route-heatmap-region-names/route.ts`, and both changed test files. **Nothing new found — converged / merge-ready.**

### Blocking
None.

### Suggestions
None.

### Nits
None worth raising.

### What I specifically re-examined this round (all clean)
- **IME-compose-then-blur** (`RegionHeatmapDetail.tsx:395,402`): the guard is correctly scoped to `onKeyDown` only. On blur during composition the browser fires `compositionend` first (flushing the composed text into `draft` via `onChange`), then `blur`, so `close(true)` commits the *completed* text — committing on blur is right, and not guarding blur is correct. No interaction with `closingRef`.
- **`closingRef` ⨉ IME**: an Enter during composition returns *before* touching `closingRef` (`:402`), so it never sets/poisons the guard; the guard is reset on every open (`:364`). Idempotent close holds.
- **maxLength vs multi-byte** (`:392`): input caps at 80 UTF-16 units; server caps `name` at 255 + trims (`route.ts:40-44`). Even all-multibyte input can't overflow the column. Picker uses the same `maxLength={80}` — consistent.
- **Layout / RTL / long names** (`:411,420,427`): editing `<input>` is `w-full max-w-[420px]` inside the `min-w-0` parent, so it shrinks rather than pushing the flex row wider; display mode uses `truncate` + `min-w-0` span. Logical-neutral utility classes flip fine in RTL. No overflow.
- **Persistence error handling**: `handleRegionRenamed` reuses `handleRegionSaved` (`FitnessHeatmapView.tsx:714`), whose reject/`saved===false`/clear-stale-error paths are already covered (view test `:372-413`). Optimistic `setRegions` keys on `entry.id` (guaranteed present via `openRegionId`) while persisting by canonical coordinate key — link unaffected, as the comment claims.
- **Clear-name round-trip**: empty commit → `onRename('')` → parent `trimmed || undefined` → `setRegions` → re-render with `value=undefined` → `useEffect` resyncs draft (editor already closed in the same `close()`). Reverts to `Map area` placeholder correctly (covered `:308-319` of the detail test).
- **Tests**: 78/78 pass; not brittle and assert the right behaviors (coordinate-key persistence, list round-trip, valid `<h2>` outline, IME no-commit, placeholder edit affordance). `yarn lint` = 0 errors, 0 warnings in any changed file.

Verified: `yarn vitest run` on both files → **78 passed**; `yarn lint` → 0 errors. No files modified.

File: lib/components/fitness/RegionHeatmapDetail.tsx [hunk 8]
```diff
@@ -321,6 +322,115 @@ const EmbedShareSection: FC<EmbedShareSectionProps> = ({
...
+  return (
+    <h2 className="truncate text-xl font-semibold tracking-tight">
+      <button
+        type="button"
+        onClick={() => setEditing(true)}
+        title="Rename"
+        className="group/name -mx-1.5 inline-flex max-w-full items-center gap-1.5 rounded-md px-1.5 py-0.5 text-left align-middle transition-colors hover:bg-accent focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring"
+      >
+        <span className="min-w-0 truncate">{value || 'Map area'}</span>
+        <Pencil className="size-3.5 shrink-0 text-muted-foreground opacity-0 transition-opacity group-hover/name:opacity-100" />
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/fitness/RegionHeatmapDetail.tsx]: Fixed in baec6b488 — added `group-focus-visible/name:opacity-100` to the pencil so it appears on keyboard focus, not just hover; keyboard/SR users now get the same edit cue when they Tab to the title. (CSS-only; `:focus-visible` opacity isn't meaningfully assertable in jsdom, so no test was added.)

**Comment by llun** [comment]
### Round 5 addressed (commit baec6b488)

- **Pencil discoverability** (@gemini-code-assist) — the rename pencil was hover-only; added `group-focus-visible/name:opacity-100` so keyboard/screen-reader users see the edit cue when the title is focused.

Thread resolved. Sub-agent round 5 reported converged/merge-ready. `yarn lint` / `yarn build` / `yarn test` (4847) green.

**Comment by llun** [comment]
## 🤖 Sub-agent review — round 6

Final convergence pass over the full `main...feat/heatmap-detail-inline-rename` diff (4 files: `RegionHeatmapDetail.tsx` + `FitnessHeatmapView.tsx` and their tests). Re-scrutinized the items called out as not-yet-nailed-down; **nothing new, nothing broken.**

**Blocking:** none.

**Suggestions:** none.

**Nits:** none.

What I verified this round:
- **Focus-visible / hover reveal** (`RegionHeatmapDetail.tsx:425,428`) — correct named-group Tailwind (`group/name` + `group-hover/name:opacity-100` / `group-focus-visible/name:opacity-100`). Keyboard users get both the button focus ring (`focus-visible:ring-2 focus-visible:ring-ring`) and the revealed pencil. No reliance on hover alone.
- **A11y of the edit affordance** — the `<button>`'s accessible name derives from its text content (`value || 'Map area'`); `title="Rename"` is a spec-correct fallback that does not override that name, and the decorative `Pencil` follows this file's established no-`aria-hidden` Lucide convention (consistent with every other icon here). Valid `<h2><button>` outline markup; input carries `aria-label="Region name"`. The role-name queries in the tests confirm the names resolve as intended.
- **Correctness** — `closingRef` idempotent close still absorbs the unmount blur (no double-commit / no cancelled-edit resurrection); the `value`→`draft` sync freezes during editing and reopens from the latest value (covered by `reflects an externally-updated name when the editor reopens`); empty/whitespace commit clears through to `null` end-to-end; world region and `onRename`-omitted callers stay static `<h2>`; the `region.id`-keyed `setRegions` update drives the detail heading with no refetch (the title reads from the memoized `openRegion`).
- **i18n** — IME-composition guard (`event.nativeEvent.isComposing`) present and tested.
- **Tests** — 78 pass across both files. `keeps an inline rename after returning to the list and reopening` is a genuine end-to-end persistence check: the `Open Renamed loop heatmap` label is derived from `region.name` in the picker row, so it only resolves if `setRegions` actually updated the list entry.
- **Lint** — `yarn lint`: 0 errors; no warnings in either changed file. Scope is correct — no new `fetch()` in a component (reuses the existing `setFitnessRouteHeatmapRegionName` client fn), no schema/`client.ts` changes, single call site, `onRename` optional so nothing else breaks.

**Converged / merge-ready.** ✅

**Comment by llun** [comment]
### ✅ Review loop converged (round 6)

Round 6 produced **no new feedback**: @gemini-code-assist reported "no review comments," and the sub-agent reviewer confirmed merge-ready with nothing new. **All review threads are resolved (0 open).**

Summary of the loop (design parity for the heatmap detail page's inline region rename):

| Round | Feedback | Fix |
|---|---|---|
| 1 | Unmount-blur race (Enter double-commit / Escape override) + allow clearing a name | Idempotent `close(save)` + reset-on-open guard; empty commit clears (commit 99777c8d6) |
| 2 | `<h2>` inside `<button>` is invalid markup | Inverted nesting to `<h2><button>…</button></h2>` (d7c486619) |
| 3 | Flex item won't truncate | `min-w-0` on the title span (07816a217) |
| 4 | IME composition: Enter/Escape mid-compose | Guard on `event.nativeEvent.isComposing` (c3656a74d) |
| 5 | Pencil cue hidden for keyboard users | `group-focus-visible/name:opacity-100` (baec6b488) |
| 6 | — | Converged |

Local verification on the final commit: `yarn lint` (0 errors) · `yarn build` (pass) · `yarn test` (549 files / 4847 tests pass). CI: Build / Lint and Prettier / CodeQL green.
